### PR TITLE
Avoid undefined behavior in shift operators during constant folding of DIExpressions.

### PR DESCRIPTION
**Pull request by tahonermann**
Bit shift operations with a shift operand greater than or equal to the bit width of the (promoted) value type result in undefined behavior according to C++ [\[expr.shift\]p1](https://eel.is/c++draft/expr.shift#1). This change adds checking for this situation and avoids attempts to constant fold `DIExpressions` that would otherwise provoke such behavior. An existing test that presumably intended to exercise shifts at the UB boundary has been updated; it now checks for shifts of 64 bits instead of 65. This issue was reported by a static analysis tool; no actual cases of shift operations that would result in undefined behavior in practice have been identified.

### CODE REVIEW

File: llvm/lib/IR/DIExpressionOptimizer.cpp [hunk 1]
```diff
@@ -59,12 +59,14 @@ foldOperationIfPossible(uint64_t Const1, uint64_t Const2,
     return Const1 - Const2;
   }
   case dwarf::DW_OP_shl: {
-    if ((uint64_t)countl_zero(Const1) < Const2)
+    if (Const2 >= std::numeric_limits<uint64_t>::digits ||
+        (uint64_t)countl_zero(Const1) < Const2)
```
Comment (tschuett) [inline llvm/lib/IR/DIExpressionOptimizer.cpp:63]: Drive-by nit: named casts are preferred, i.e., static_cast.

Comment (tahonermann) [reply llvm/lib/IR/DIExpressionOptimizer.cpp]: Fixed, thanks!

**Review by rastogishubham — APPROVED** [review]
LGTM thanks for catching the bug!
